### ISSUE TITLE
Client remains disconnected when network connection is restored

### ISSUE THREAD

**Issue by nick7ikin** [hazelcast/hazelcast#6872]
I'm investigating client reconnection issue. I can't reproduce it 100%, it happens sometimes. I'm using LifeCycleListener to detect client connection state. Steps are following:
1. Hazelcast client connects to cluster member
2. I turn off networking (on Ubuntu system tray -> uncheck Enable Networking) on client side
3. client hardbeat detects connection issue
4. after some period of time I turn on networking back
5. client restores connection for short period of time and then loses it again.

Hazelcast logs.
When connection is lost hazelcast prints:
```
Nov 25, 2015 1:25:48 PM com.hazelcast.client.connection.nio.ClientConnection
WARNING: Connection [/192.168.0.22:5701] lost. Reason: Socket explicitly closed
Nov 25, 2015 1:25:48 PM com.hazelcast.core.LifecycleService
INFO: HazelcastClient[hz.client_0_mememe][3.5.3] is CLIENT_DISCONNECTED
Nov 25, 2015 1:25:48 PM com.hazelcast.client.spi.impl.ClusterListenerSupport
WARNING: Unable to get alive cluster connection, try in 4999 ms later, attempt 1 of 10000.
Nov 25, 2015 1:25:53 PM com.hazelcast.client.spi.ClientInvocationService
WARNING: There are packets which are not processed 3
Nov 25, 2015 1:25:53 PM com.hazelcast.client.spi.impl.ClusterListenerSupport
WARNING: Unable to get alive cluster connection, try in 4998 ms later, attempt 2 of 10000.
Nov 25, 2015 1:25:58 PM com.hazelcast.client.spi.impl.ClusterListenerSupport
WARNING: Unable to get alive cluster connection, try in 4999 ms later, attempt 3 of 10000.
```

when connection is restored
```
Nov 25, 2015 1:26:03 PM com.hazelcast.core.LifecycleService
INFO: HazelcastClient[hz.client_0_mememe][3.5.3] is CLIENT_CONNECTED

Nov 25, 2015 1:26:08 PM com.hazelcast.client.connection.nio.ClientConnection
WARNING: Connection [/192.168.0.22:5701] lost. Reason: Socket explicitly closed
Nov 25, 2015 1:26:08 PM com.hazelcast.client.connection.nio.ClientReadHandler
WARNING: hz.client_0_mememe.ClientInSelector Closing socket to endpoint Address[192.168.0.22]:5701, Cause:java.io.EOFException: Remote socket closed!
Nov 25, 2015 1:26:08 PM com.hazelcast.core.LifecycleService
INFO: HazelcastClient[hz.client_0_mememe][3.5.3] is CLIENT_CONNECTED
Nov 25, 2015 1:26:08 PM com.hazelcast.core.LifecycleService
INFO: HazelcastClient[hz.client_0_mememe][3.5.3] is CLIENT_DISCONNECTED

Nov 25, 2015 1:26:13 PM com.hazelcast.client.spi.ClientInvocationService
WARNING: There are packets which are not processed 1
```

when the issue isn't reproduced the logs are:
```
Nov 25, 2015 12:02:40 PM com.hazelcast.client.spi.ClientInvocationService
WARNING: There are packets which are not processed 1
Nov 25, 2015 12:02:47 PM com.hazelcast.client.spi.impl.ClusterListenerSupport
WARNING: Unable to get alive cluster connection, try in 0 ms later, attempt 1 of 10000.

Nov 25, 2015 12:02:59 PM com.hazelcast.client.spi.impl.ClusterListenerSupport
WARNING: Unable to get alive cluster connection, try in 0 ms later, attempt 2 of 10000.
Nov 25, 2015 12:03:08 PM com.hazelcast.core.LifecycleService
INFO: HazelcastClient[hz.client_0_mememe3322][3.5.3] is CLIENT_CONNECTED

Nov 25, 2015 12:03:20 PM com.hazelcast.client.connection.nio.ClientConnection
WARNING: Connection [/192.168.1.100:5701] lost. Reason: Socket explicitly closed
Nov 25, 2015 12:03:20 PM com.hazelcast.client.connection.nio.ClientReadHandler
WARNING: hz.client_0_mememe3322.ClientInSelector Closing socket to endpoint Address[192.168.1.100]:5701, Cause:java.io.EOFException: Remote socket closed!
Nov 25, 2015 12:03:20 PM com.hazelcast.core.LifecycleService
INFO: HazelcastClient[hz.client_0_mememe3322][3.5.3] is CLIENT_DISCONNECTED
Nov 25, 2015 12:03:20 PM com.hazelcast.core.LifecycleService
INFO: HazelcastClient[hz.client_0_mememe3322][3.5.3] is CLIENT_CONNECTED

Nov 25, 2015 12:03:25 PM com.hazelcast.client.spi.ClientInvocationService
WARNING: There are packets which are not processed 1
```
So it looks very similar except the order of last CLIENT_DISCONNECTED CLIENT_CONNECTED messages. I believe there could racing between different types of timeouts.

Configuration ```hazelcast-client.xml```
```
    <network>
        <smart-routing>false</smart-routing>
        <redo-operation>true</redo-operation>
        <connection-attempt-limit>10000</connection-attempt-limit>
        <connection-attempt-period>5000</connection-attempt-period>
        <connection-timeout>3000</connection-timeout>
    </network>
```
also several options are configured in runtime:
```
clientConfig.setProperty("hazelcast.client.heartbeat.interval", "1000");     
clientConfig.setProperty("hazelcast.client.max.no.heartbeat.seconds", "2");  
clientConfig.setProperty("hazelcast.heartbeat.interval.seconds", "1");       
clientConfig.setProperty("hazelcast.client.invocation.timeout.seconds", "1");
clientConfig.setProperty("hazelcast.client.heartbeat.timeout", "5000");      
```

Could you share your ideas how to overcome this issue?

**Comment by sancar** [comment]
Hi @nick7ikin 

These two properties belongs to node not client. These do not affect the client.

        clientConfig.setProperty("hazelcast.client.max.no.heartbeat.seconds", "2");
        clientConfig.setProperty("hazelcast.heartbeat.interval.seconds", "1");

From experience, sometimes connection timeout 3000 (3 seconds) is not enough if you are connecting to a server in different location.

Other than that I tried with your configuration several times , it  does not reproduce on my local machine. I don't see the CLIENT_DISCONNECTED after client connected back.

How many nodes are you trying with ?
Are you setting the two properties I mentioned above to server side also ?

**Comment by nick7ikin** [comment]
Hi @sancar 

Thanks for the note about server-side properties.

I tried with two nodes (server and client) connected with local network. I could catch this issue when my client was frequently disconnected. Currently I've increased ```client.heartbeat.timeout``` according to #6873 and now it is several times greater than ```heartbeat.interval.seconds```. That's why I can't reproduce it also. Thanks for your help.
